### PR TITLE
Manual merge 3.x into master

### PR DESCRIPTION
**Pull request by wbloszyk**

### CODE REVIEW

File: docs/reference/advanced_configuration.rst [hunk 1]
```diff
@@ -8,10 +8,10 @@ Full configuration options:
     sonata_media:
         db_driver: doctrine_orm
         class:
-            media:              Application\Sonata\MediaBundle\Entity\Media
-            gallery:            Application\Sonata\MediaBundle\Entity\Gallery
-            gallery_item:       Application\Sonata\MediaBundle\Entity\GalleryItem
-            category:           Application\Sonata\ClassificationBundle\Entity\Category
+            media: App\Entity\SonataMediaMedia
+            gallery: App\Entity\SonataMediaGallery
+            gallery_item: App\Entity\GalleryItem
```
Comment (jordisala1991) [inline docs/reference/advanced_configuration.rst:13]: SonataMediaGalleryItem

File: docs/reference/installation.rst [hunk 2]
```diff
@@ -1,157 +1,105 @@
...
+
+.. code-block:: yaml
+
+    # config/packages/sonata_media.yaml
+
+    sonata_media:
+        class:
+            media: App\Entity\SonataMediaMedia
+            gallery: App\Entity\SonataMediaGallery
+            gallery_has_media: App\Entity\SonataMediaGalleryHasMedia
```
Comment (jordisala1991) [inline docs/reference/installation.rst:82]: `gallery_item: App\Entity\SonataMediaGalleryItem`

File: docs/reference/installation.rst [hunk 3]
```diff
@@ -161,104 +109,274 @@ Once you have done that, you can configure the Media bundle itself:
...
+    {
+        /**
+         * @ORM\Id
+         * @ORM\GeneratedValue
+         * @ORM\Column(type="integer")
+         */
+        protected $id;
+    }
+
+and ``src/Entity/SonataMediaGalleryHasMedia``::
```
Comment (jordisala1991) [inline docs/reference/installation.rst:183]: `SonataMediaGalleryItem`

File: docs/reference/installation.rst [hunk 4]
```diff
@@ -161,104 +109,274 @@ Once you have done that, you can configure the Media bundle itself:
...
+
+    // src/Entity/SonataMediaGalleryHasMedia.php
+
+    use Doctrine\ORM\Mapping as ORM;
+    use Sonata\MediaBundle\Entity\BaseGalleryHasMedia;
```
Comment (jordisala1991) [inline docs/reference/installation.rst:188]: `BaseGalleryItem`

File: docs/reference/installation.rst [hunk 5]
```diff
@@ -161,104 +109,274 @@ Once you have done that, you can configure the Media bundle itself:
...
+
+    /**
+     * @ORM\Entity
+     * @ORM\Table(name="media__gallery_has_media")
```
Comment (jordisala1991) [inline docs/reference/installation.rst:192]: media__gallery_item

File: docs/reference/installation.rst [hunk 6]
```diff
@@ -161,104 +109,274 @@ Once you have done that, you can configure the Media bundle itself:
...
+         * @ORM\Column(type="integer")
+         */
+        protected $id;
+    }
+
+The only thing left is to update your schema::
+
+    bin/console doctrine:schema:update --force
+
+Doctrine PHPCR Configuration
```
Comment (jordisala1991) [inline docs/reference/installation.rst:208]: Same for the other persistence bundles

File: src/DependencyInjection/SonataMediaExtension.php [hunk 7]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
     }
+
+    private function registerSonataDoctrineMapping(array $config): void
+    {
+        $collector = DoctrineCollector::getInstance();
+
+        $collector->addAssociation(
+            $config['class']['media'],
+            'mapOneToMany',
+            OptionsBuilder::createOneToMany('galleryHasMedias', $config['class']['gallery_has_media'])
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:564]: `galleryItems` and `gallery_item`

File: src/DependencyInjection/SonataMediaExtension.php [hunk 8]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
+                ->cascade(['persist'])
+                ->mappedBy('media')
+        );
+
+        $collector->addAssociation(
+            $config['class']['gallery_has_media'],
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:570]: `gallery_item`

File: src/DependencyInjection/SonataMediaExtension.php [hunk 9]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
+            'mapManyToOne',
+            OptionsBuilder::createManyToOne('gallery', $config['class']['gallery'])
+                ->cascade(['persist'])
+                ->inversedBy('galleryHasMedias')
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:574]: `galleryItems`

File: src/DependencyInjection/SonataMediaExtension.php [hunk 10]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
+                ->addJoin([
+                    'name' => 'gallery_id',
+                    'referencedColumnName' => 'id',
+                    'onDelete' => 'CASCADE',
+                ])
+        );
+
+        $collector->addAssociation(
+            $config['class']['gallery_has_media'],
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:583]: `gallery_item`

File: src/DependencyInjection/SonataMediaExtension.php [hunk 11]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
+            'mapManyToOne',
+            OptionsBuilder::createManyToOne('media', $config['class']['media'])
+                ->cascade(['persist'])
+                ->inversedBy('galleryHasMedias')
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:587]: `galleryItems`

File: src/DependencyInjection/SonataMediaExtension.php [hunk 12]
```diff
@@ -541,4 +553,67 @@ private function configureResizers(ContainerBuilder $container, array $config):
...
+                    'name' => 'media_id',
+                    'referencedColumnName' => 'id',
+                    'onDelete' => 'CASCADE',
+                ])
+        );
+
+        $collector->addAssociation(
+            $config['class']['gallery'],
+            'mapOneToMany',
+            OptionsBuilder::createOneToMany('galleryHasMedias', $config['class']['gallery_has_media'])
```
Comment (jordisala1991) [inline src/DependencyInjection/SonataMediaExtension.php:598]: same as above

File: tests/App/Entity/GalleryHasMedia.php [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+namespace Sonata\MediaBundle\Tests\App\Entity;
+
+use Doctrine\ORM\Mapping as ORM;
+use Sonata\MediaBundle\Entity\BaseGalleryHasMedia;
+
+/**
+ * @ORM\Entity
+ * @ORM\Table(name="media__gallery_has_media")
+ */
+class GalleryHasMedia extends BaseGalleryHasMedia
```
Comment (jordisala1991) [inline tests/App/Entity/GalleryHasMedia.php:23]: GalleryItem. This class needs to be renamed, and fixed for all the gallery_had_media mentions

File: tests/App/config.yml [hunk 14]
```diff
@@ -15,6 +16,9 @@ sonata_media:
     db_driver: doctrine_orm
     class:
         media: Sonata\MediaBundle\Tests\App\Entity\Media
+        gallery: Sonata\MediaBundle\Tests\App\Entity\Gallery
+        gallery_has_media: Sonata\MediaBundle\Tests\App\Entity\GalleryHasMedia
```
Comment (jordisala1991) [inline tests/App/config.yml:20]: gallery_item

File: tests/Functional/RoutingTest.php [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public function getRoutes(): iterable
+    {
+        // API
+        yield ['nelmio_api_doc_index', '/api/doc/{view}', ['GET']];
+        yield ['sonata_api_media_gallery_get_galleries', '/api/media/galleries.{_format}', ['GET']];
+        yield ['sonata_api_media_gallery_get_gallery', '/api/media/galleries/{id}.{_format}', ['GET']];
+        yield ['sonata_api_media_gallery_get_gallery_medias', '/api/media/galleries/{id}/medias.{_format}', ['GET']];
+        yield ['sonata_api_media_gallery_get_gallery_galleryhasmedias', '/api/media/galleries/{id}/galleryhasmedias.{_format}', ['GET']];
```
Comment (jordisala1991) [inline tests/Functional/RoutingTest.php:85]: gallery_item

**Comment by wbloszyk** [comment]
For now I merge 3.x into master. Before we merge it we should consider fix two issue:
- change all `GalleryHasMedia` to `GalleryItem` becouse it is incompleted in master branch
- (optional) remove `{@inheritdoc}` in master (this will remove it from new code and make merge simpler)

**Comment by jordisala1991** [comment]
The incomplete part of sonata gallery item comes from this Pr , it was complete on master

**Comment by wbloszyk** [comment]
@jordisala1991 I know your comments come from 3.x. But in master are also incomplete part. Just look at:
- https://github.com/sonata-project/SonataMediaBundle/pull/1777
- https://github.com/sonata-project/SonataMediaBundle/pull/1776

**Comment by jordisala1991** [comment]
Agree on that PRs but my comments has to be adressed here.

**Comment by wbloszyk** [comment]
@jordisala1991 Finnaly this merge is RTM. Can you check it? 

BTW
It is my first merge PR with this kind of changes. I hope I can make this change in other commit(but in the same PR). Some change(which is not in conflict) didn't commit after `merge --continue`. It is correct or I did something wrong?

**Comment by wbloszyk** [comment]
Can we merge it?

**Pull request merged**